### ISSUE TITLE
Platform specific files are not resolved when using `metro-serializer-esbuild`

### ISSUE THREAD

**Issue by renchap** [microsoft/rnx-kit#1864]
### What happened?

I am trying to enable tree-shaking on our app, which is already using `metro-config` and `babel-preset-metro-react-native`. It builds correctly with this setup.

I tried two different ways:
- enabling `metro-serializer-esbuild` in Metro config as specified in the docs
- switching to `@rnx-kit/cli` and enabling tree-shaking

Both ended up with the same errors:
```
yarn exec react-native rnx-bundle --entry-file index.js --platform ios --dev false
…
error Build failed with 75 errors:
../../node_modules/@invertase/react-native-apple-authentication/lib/index.js:20:50: ERROR: Could not resolve "./AppleButton"
../../node_modules/react-native/Libraries/Alert/Alert.js:7:47: ERROR: Could not resolve "../Utilities/Platform"
../../node_modules/react-native/Libraries/Alert/Alert.js:9:54: ERROR: Could not resolve "./RCTAlertManager"
../../node_modules/react-native/Libraries/Animated/Animated.js:3:47: ERROR: Could not resolve "../Utilities/Platform"
../../node_modules/react-native/Libraries/Animated/NativeAnimatedHelper.js:9:47: ERROR: Could not resolve "../Utilities/Platform"
...
```

This seems to happen for all files with a platform specific extension (`.ios.js` or `.android.js`).

I tried disabling various babel plugins (like `babel-plugin-module-resolver`) but nothing fixed it. We are running RN 0.70.0-rc4 in a yarn 3 monorepo.

### Affected Package

@rnx-kit/metro-serializer-esbuild

### Version

1.0.11

### Which platforms are you seeing this issue on?

- [X] Android
- [X] iOS
- [ ] macOS
- [ ] Windows

### System Information

```true
yarn exec react-native info
info Fetching system and libraries information...
System:
    OS: macOS 12.5.1
    CPU: (10) arm64 Apple M1 Max
    Memory: 3.36 GB / 32.00 GB
    Shell: 5.8.1 - /bin/zsh
  Binaries:
    Node: 16.15.1 - /private/var/folders/cn/qgvjcjx97mj8rc2vx37vgkvh0000gn/T/xfs-1a5d27f9/node
    Yarn: 3.2.0 - /private/var/folders/cn/qgvjcjx97mj8rc2vx37vgkvh0000gn/T/xfs-1a5d27f9/yarn
    npm: 8.11.0 - ~/.nvm/versions/node/v16.15.1/bin/npm
    Watchman: 2022.08.15.00 - /opt/homebrew/bin/watchman
  Managers:
    CocoaPods: 1.11.3 - /var/folders/cn/qgvjcjx97mj8rc2vx37vgkvh0000gn/T/frum_53057_1661525526964/bin/pod
  SDKs:
    iOS SDK:
      Platforms: DriverKit 21.4, iOS 15.5, macOS 12.3, tvOS 15.4, watchOS 8.5
    Android SDK:
      API Levels: 30, 31, 32, 33
      Build Tools: 30.0.2, 30.0.3, 31.0.0, 32.0.0, 32.1.0, 33.0.0
      System Images: android-24 | Google APIs ARM 64 v8a, android-25 | Google APIs ARM 64 v8a, android-26 | Google APIs ARM 64 v8a, android-28 | Google ARM64-V8a Play ARM 64 v8a, android-32 | Google APIs ARM 64 v8a
      Android NDK: Not Found
  IDEs:
    Android Studio: 2021.2 AI-212.5712.43.2112.8609683
    Xcode: 13.4.1/13F100 - /usr/bin/xcodebuild
  Languages:
    Java: 17.0.4 - /usr/bin/javac
  npmPackages:
    @react-native-community/cli: Not Found
    react: Not Found
    react-native: Not Found
    react-native-macos: Not Found
  npmGlobalPackages:
    *react-native*: Not Found
```

### Steps to Reproduce

- Enable tree-shaking in `package.json`
- Create a bundle with `rnx-bundle`

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by renchap** [comment]
When enabling eslint debug output, I get this:

```
● [DEBUG] Resolving import "../../Utilities/Platform" in directory "/Users/renchap/dev/notos/node_modules/react-native/Libraries/Components/ScrollView" of type "require-call"

  Read 11 entries for directory "/Users/renchap/dev/notos/node_modules/react-native/Libraries/Components/ScrollView"
  Read 57 entries for directory "/Users/renchap/dev/notos/node_modules/react-native/Libraries/Utilities"
  No "browser" map found in directory "/Users/renchap/dev/notos/node_modules/react-native/Libraries/Utilities"
  Attempting to load "/Users/renchap/dev/notos/node_modules/react-native/Libraries/Utilities/Platform" as a file
    Checking for file "Platform"
    Checking for file "Platform.tsx"
    Checking for file "Platform.ts"
    Checking for file "Platform.jsx"
    Checking for file "Platform.js"
    Checking for file "Platform.css"
    Checking for file "Platform.json"
    Failed to find file "Platform"
  Attempting to load "/Users/renchap/dev/notos/node_modules/react-native/Libraries/Utilities/Platform" as a directory
    Failed to read directory "/Users/renchap/dev/notos/node_modules/react-native/Libraries/Utilities/Platform"
```

It is not checking for platform extensions, I am not sure why.

**Comment by tido64** [comment]
Hi @renchap, sorry you're seeing all these issues.

1. What does your `metro.config.js` look like? These don't look like errors coming from our serializer. Serialization is the last step in the bundling process, when you are about to write to disk. Module resolution and transpilation is still all Metro.
2. Do you have a repro somewhere we can look at?
3. Where does the output you pasted come from? If it's from ESLint, it's got nothing to do with Metro or the serializer.

**Comment by renchap** [comment]
Thanks for your quick answer @tido64

I edited my previous comment, I means `esbuild`, not :eslint` 🤦 

Here is my Metro config:

```ts
const { makeMetroConfig } = require("@rnx-kit/metro-config")
const MetroSymlinksResolver = require("@rnx-kit/metro-resolver-symlinks")
const {
  MetroSerializer,
  esbuildTransformerConfig,
} = require("@rnx-kit/metro-serializer-esbuild")
const { getDefaultConfig } = require("metro-config")

module.exports = (async () => {
  const {
    resolver: { sourceExts, assetExts },
  } = await getDefaultConfig()

  return makeMetroConfig({
    resolver: {
      resolveRequest: MetroSymlinksResolver({
        // From https://github.com/microsoft/rnx-kit/issues/1804
        // Otherwise it does not find our RN packages in the monorepo
        remapModule: (_context, moduleName, _platform) => {
          if (moduleName.match(/(^react-native$)|(^react-native\/)/)) {
            return require.resolve(moduleName)
          }
          return moduleName
        },
      }),
      assetExts: assetExts.filter((ext) => ext !== "svg"),
      sourceExts: [...sourceExts, "mjs", "svg"],
    },
    serializer: {
      customSerializer: MetroSerializer([],
        { fabric: false, logLevel: "debug" },
      ),
    },
    transformer: {
      ...esbuildTransformerConfig,
      babelTransformerPath: require.resolve("react-native-svg-transformer"),
    },
  })
})()
```

Now that I think about it, this may be caused by the `remapModule` used to solve the same issue as #1804…

**Comment by tido64** [comment]
That `remapModule` probably shouldn't be there. What happens if you take it out?

**Comment by tido64** [comment]
Did you also try resetting the cache with `--reset-cache`? This is very strange. esbuild is not supposed to resolve anything as we're feeding it the graph output from Metro. Do you have a repro I can take a look at?

**Comment by renchap** [comment]
Yes I get these errors when running `yarn exec react-native bundle --entry-file index.js --platform ios --dev false --verbose --bundle-output index.ios.jsbundle --reset-cache`, so no cache.

I can work on a repro, but this will be next week. I assume this must come from something trivial somewhere in my setup, but I have not been able to find out what after some hours…

**Comment by renchap** [comment]
I started commenting and removing all my App's code, then config, and was not able to reproduce… until I remembered I am running the latest 0.70 RC, and this is the issue!

Here is a repro: https://github.com/renchap/rnxkit-repro

You can run:
```
yarn
yarn bundle # works
yarn bundle-esbuild # fails
```

If you edit the `package.json` to go back to version 0.69.5, it bundles fine with `esbuild`.

Is this expected?

**Comment by tido64** [comment]
> Here is a repro: https://github.com/renchap/rnxkit-repro

Thanks for creating the repro! It looks like Metro's graph format recently changed, causing all the lookups to fail. I will try to get this fixed ASAP.

**Comment by tido64** [comment]
@renchap: Would you mind testing the fix 👆 and let me know how it goes?

**Comment by renchap** [comment]
I had a bit of trouble figuring out how to use my local copy of the package in my repo, but I managed to do it and I can confirm it works with the newly released RN 0.70.0 🎉

**Comment by tido64** [comment]
This should now be fixed in [0.1.14](https://github.com/microsoft/rnx-kit/releases/tag/%40rnx-kit%2Fmetro-serializer-esbuild%400.1.14). Let us know how it goes. Thanks again for filing this issue!
